File: src/main/java/qj/codejam/y2008/qualification/b/TrainTimetable.java
package qj.codejam.y2008.qualification.b;

import qj.codejam.CodeJam;
import qj.util.Cols;
import qj.util.DateUtil;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;

public class TrainTimetable {
	public static void main(String[] args) {
		String resName = "sample.in";
//		String resName = "B-small-practice.in";
//		String resName = "B-large-practice.in";
		CodeJam.solver(resName, lineF -> {
			long turnTime = Integer.parseInt(lineF.e()) * DateUtil.MINUTE_LENGTH;
			String[] split = lineF.e().split(" ");
			int na = Integer.parseInt(split[0]);
			int nb = Integer.parseInt(split[1]);

			SimpleDateFormat df = new SimpleDateFormat("HH:mm");
			final List<Long> needAtA = new ArrayList<>();
			final List<Long> readyAtB = new ArrayList<>();
			final List<Long> needAtB = new ArrayList<>();
			final List<Long> readyAtA = new ArrayList<>();
			for (int i = 0; i < na; i++) {
				String[] timeSplit = lineF.e().split(" ");
				needAtA.add(DateUtil.parse(timeSplit[0], df).getTime());
				readyAtB.add(DateUtil.parse(timeSplit[1], df).getTime() + turnTime);
			}
			for (int i = 0; i < nb; i++) {
				String[] timeSplit = lineF.e().split(" ");
				needAtB.add(DateUtil.parse(timeSplit[0], df).getTime());
				readyAtA.add(DateUtil.parse(timeSplit[1], df).getTime() + turnTime);
			}

			return () -> solve(readyAtA, needAtA) + " " + solve(readyAtB, needAtB);
		})
		.solve();
	}

	private static int solve(List<Long> readys, List<Long> needs) {
		if (needs.size() == 0) {
			return 0;
		}
		
		readys = Cols.reverse(Cols.sort(readys));
		needs = Cols.reverse(Cols.sort(needs));
		
		for (Long ready : readys) {
			Long need = needs.get(0);
			if (ready <= need) {
				needs.remove(0);
				if (needs.size() == 0) {
					return 0;
				}
			}
		}
		
		return needs.size();
	}
}

File: src/main/scala/qs/codejam/y2008/qualification/b/TrainTimetable.scala
package qs.codejam.y2008.qualification.b

import qs.codejam.CodeJam
import scala.collection.mutable.ListBuffer
import java.text.SimpleDateFormat

object TrainTimetable {
  def main(args: Array[String]) {
    val resName = "sample.in"

    CodeJam.solver(resName, readAndSolve = (lineF) => {
      val turnTime: Long = lineF().toInt * 1000 * 60
      val split = lineF().split(" ")
      val na: Int = split(0).toInt
      val nb: Int = split(1).toInt

      val needAtA = new ListBuffer[Long]
      val readyAtB = new ListBuffer[Long]
      val needAtB = new ListBuffer[Long]
      val readyAtA = new ListBuffer[Long]

      val df = new SimpleDateFormat("HH:mm")

      for (i <- 1 to na) {
        val timeSplit = lineF().split(" ")

        needAtA += df.parse(timeSplit(0)).getTime
        readyAtB += df.parse(timeSplit(1)).getTime + turnTime
      }
      for (i <- 1 to nb) {
        val timeSplit = lineF().split(" ")

        needAtB += df.parse(timeSplit(0)).getTime
        readyAtA += df.parse(timeSplit(1)).getTime + turnTime
      }

      () => solve(readyAtA.result(), needAtA.result()) + " " + solve(readyAtB.result(), needAtB.result())
    })
    .solve()
  }

  private def solve(readys: List[Long], needs: List[Long]): Int = {
    if (needs.size == 0) {
      return 0
    }

    val ry = readys.sorted.reverse
    var nd = needs.sorted.reverse

    for (ready <- ry) {
      val need = nd.head
      if (ready <= need) {
        nd = nd.tail
        if (nd.size == 0) {
          return 0
        }
      }
    }
    nd.size
  }
}